man/Rit.1

.\" Manpage for Rit.
.\" Contact <moeinn.com@gmail.com> to correct errors or typos.

.TH Rit "1" "Spring 2020" "The Rit Project" "Rit man page" 

.SH "NAME"
Rit \- Rename it ! :)

.SH SYNOPSIS

\&\fBRit\fR [\fI\s-1FILE\s0\fR]...[\fI\s-1OPTION\s0\fR]  
.br
\n\&\fBRit\fR [\fI\s-1OPTION\s0\fR] [\fI\s-1FILE\s0\fR]  
.br
\&\fBRit\fR [\fI\s-1FILE \s0\fR] [\fI\s-1%format abbriviation\s0\fR] [\fI\s-1OPTION\s0\fR]  

.SH DESCRIPTION
.PP
Rit rename photo or multiple photo(s) to the Time as were taken .
.br
For example you are on vacation and you took more than 200 
pictures , some pictures are taken by canon camera and the 
rest of taken by your iphone and after the vacation you 
transfer data to your laptop or PC then you seen every device 
have own type of name format so when you move pictures in 1 
folder you see some dirty :( , Rit rename it by first pic and 
with your favourite time format =))
.PP
When you dont set time format , default time format is : 
.PP
        \&\fB%F ( %Y-%m-%d ) : YYYY-MM-DD\fR

.TP
\fB\-f\fR, \fB\-\-format\fR
FORMAT controls the output . see Examples more useful .
.TP
\fB\-j\fR, \fB\-\-jformat\fR
jalali format 
.TP
\fB\-r\fR, \fB\-\-recursive\fR
rename directories recursively
.TP 
\fB\-v\fR, \fB\-\-verbos\fR
explain what is being done 
.PP
\fB\Interpreted sequences are :\fR
.TP
%a
locale's abbreviated weekday name (e.g., Sun)
.TP
%A
locale's full weekday name (e.g., Sunday)
.TP
%b
locale's abbreviated month name (e.g., Jan)
.TP
%B
locale's full month name (e.g., January)
.TP
%c
locale's date and time (e.g., Thu Mar  3 23:05:25 2005)
.TP
%C
century; like %Y, except omit last two digits (e.g., 20)
.TP
%d
day of month (e.g., 01)
.TP
%e
day of month, space padded; same as %_d
.TP
%F
full date; same as %Y\-%m\-%d
.TP
%g
last two digits of year of ISO week number (see %G)
.TP
%G
year of ISO week number (see %V); normally useful only with %V
.TP
%h
same as %b
.TP
%H
hour (00..23)
.TP
%I
hour (01..12)
.TP
%j
day of year (001..366)
.TP
%k
hour, space padded ( 0..23); same as %_H
.TP
%l
hour, space padded ( 1..12); same as %_I
.TP
%m
month (01..12)
.TP
%M
minute (00..59)
.TP
%N
nanoseconds (000000000..999999999) #TODO Fix it
.TP
%p
locale's equivalent of either AM or PM; blank if not known
.TP
%P
like %p, but lower case
.TP
%q
quarter of year (1..4)
.TP
%r
locale's 12\-hour clock time (e.g., 11:11:04 PM)
.TP
%R
24\-hour hour and minute; same as %H:%M
.TP
%s
seconds since 1970\-01\-01 00:00:00 UTC
.TP
%S
second (00..60)
.TP
%T
time; same as %H:%M:%S
.TP
%u
day of week (1..7); 1 is Monday
.TP
%U
week number of year, with Sunday as first day of week (00..53)
.TP
%V
ISO week number, with Monday as first day of week (01..53)
.TP
%w
day of week (0..6); 0 is Sunday
.TP
%W
week number of year, with Monday as first day of week (00..53)
.TP
%X
locale's time representation (e.g., 23:13:48)
.TP
%y
last two digits of year (00..99)
.TP
%Y
year
.TP
%Z
alphabetic time zone abbreviation (e.g., EDT)
.PP
By default, date pads numeric fields with zeroes.
The following optional flags may follow '%':
.TP
\-
(hyphen) do not pad the field
.TP
_
(underscore) pad with spaces
.TP
0
(zero) pad with zeros
.TP
^
use upper case if possible
.TP
#
.PP
\fB\Jalali time format :\fR
.TP
%a
locale's abbreviated weekday name (e.g., Yek)
.TP
%A
locale's full weekday name (e.g., Yekshanbe)
.TP
%b
locale's abbreviated month name (e.g., Far)
.TP
%B
locale's full month name (e.g., Farvardin)
.TP
%c
locale's date and time (e.g., Yekshanbeh 20 Ordibehesht 1399 23:05:25 )
.TP
%d
day of month (e.g, 01)
.TP
%H 
hour (00..23)
.TP
%I
hour (01..12)
.TP
%j
day of year (001..366)
.TP
%m
month (01..12)
.TP
%M 
minute (00..59)
.TP
%p 
locale's equivalent of either AM or PM; blank if not known
.TP
%S
second (00..60)
.TP
%U
week number of year, with Shanbeh as first day of week (00..53)
.TP
%w
day of week (0..6); 1 is Yekshanbeh
.TP
%W
week number of year, with Doshanbeh as first day of week (00..53)
.TP
%X     
locale's time representation (e.g., 23:13:48)
.TP
%Y
year

.TP
\fB\-\-help\fR
.br
show short help 
.TP
\fB\-\-version\fR
version information and exit
.SH EXAMPLES
.PP
Rename simply with default time format :    

        \&\fBRit photo.jpg photo2.jpg\fR
.br
output :
        \&\fB"2020-02-11.jpg" and "2020-03-25.jpg" \fR

Rename with use strftime and see what change verbosity :

        \&\fBRit Photos/* %c -fv\fR
.br
output :
        \&\fB"Wed Mar 20 08:41:14 2019.jpg"\fR

Rename with strftime and add a name inside it :

        \&\fBRit home/moein/vacation/* Abhar\\ %Y-%b-%d -fv\fR
.br        
output :
        \&\fB"Abhar 2020-Feb-14.jpg"\fR
.PP
Rename with jalali time :
        
        \&\fBRit photo.jpg photo2.jpg -j\fR
.br
output :
        \&\fB"1395-04-31.jpg" and "1399-12-01.jpg"\fR

.PP       
Rename with jalali time and see what change verbosity :

        \&\fBRit dog.jpg %c -jv\fR
.br
output :
        \&\fB"Yekshanbeh 20 Ordibehesht 1399 23:12:25.jpg"\fR
.PP       
Rename directories recursively :

        \&\fBRit -r ~/Pictures \fR
.PP
You can rename like : 

        \&\fBfind . -type f -iname "2020*jpg" | xargs -IX Rit X %c -f\fR
.br
or :
        \&\fBls *jpeg | xargs -IX Rit X %B -j\fR
.SH AUTHOR
Written by \(co Moein Halvaei <moeinn.com@gmail.com> 
.br
send me your idea and suggestion 
about this Project that makes me happy : )
.SH BUGS
 Still fixing :D .
.br
.SH SEE ALSO
mv(1) rename(2)
   


Rit/Rit.py
#!/usr/bin/python3

import os
import re
import sys
import time
import argparse
import PIL.Image
import PIL.ExifTags
from PIL import Image
from datetime import datetime
from Rit.jalali import JalaliFormat

class Rename :

    def __init__ (self):
        #Defualt time format
        self.format = '%Y-%m-%d'
        self.jalali = False 
        self.Verbos = False
        self.recursive = False
        self.fullpath = '' 
        self.count = 1
        self.cnt = 0
        self.path = []
        self.shit = []
        self.args = {}
        self.dic = {}
        self.progress = 0.0

    #read Modified DateTime from exif data 
    def readMeta(self,f) :
        img = PIL.Image.open(f)

        #some photos are damaged and we got error
        #so use exception

        try:
            exif = {
                PIL.ExifTags.TAGS[k]: v
                for k, v in img._getexif().items()
                if k in PIL.ExifTags.TAGS
            }

        except :
            if self.Verbos :
                print(' Rit : Oops no meta :]')

        try :
            realTime = exif['DateTime']
            #because strptime has no str obj ...
            image_time = datetime.strptime(realTime, "%Y:%m:%d %H:%M:%S")
            new = image_time.strftime(self.format)
            alldate = image_time.strftime("%Y%m%d%H%M%S")

            if self.jalali :
                alldate = JalaliFormat(image_time,self.format).jsort()
                new = JalaliFormat(image_time,self.format).jconvert()

        except :
            realTime = time.strftime("%Y:%m:%d %H:%M:%S",time.gmtime(os.path.getmtime(f)))
            image_time = datetime.strptime(realTime, "%Y:%m:%d %H:%M:%S")
            new = image_time.strftime(self.format)
            alldate = image_time.strftime("%Y%m%d%H%M%S")

            if self.jalali :
                alldate = JalaliFormat(image_time,self.format).jsort()
                new = JalaliFormat(image_time,self.format).jconvert()
        
        times = new , alldate
        return times

    def SortTodic(self,things):
        
        # find full path of file (absolutized path)
        self.fullpath = os.path.abspath(things)

        #Sort file by erlier Time
        Priority = self.readMeta(things)
        self.dic[Priority[1]] = self.fullpath ,Priority[0]

    def Rit (self):

        if not self.path and self.args['help'] is not True and self.args['version'] is not True :
            print (' Rit: missing operand \nTry \'Rit --help\'')
        else :
                for files in self.path :
                    self.progress += 1.0 / len(self.path)

                    try :
                        if os.path.isdir(files) == False:
                            self.SortTodic(files) 
                        else:
                            
                            if self.recursive == False :
                                print(" Rit : \'%s\' is directory ; need -r "%(files.split('/')[-1]) )
                            
                            else :
                                start = os.path.expanduser(files)

                                #walking...
                                for root, dirs, Files in os.walk(start, topdown = False):
                                    for name in Files:
                                        thing = os.path.join(root, name)

                                        #check if file image or not
                                        try :
                                            img = Image.open(thing)
                                            self.SortTodic(thing)
                                        except :
                                            if self.Verbos :
                                                print('not image')
                                self.shit.append(files)
                            
                    except OSError :
                                print (
                                        ' Rit: cannot stat \'%s\': '%files
                                        +'No such file or directory'
                                ) 

                                self.shit.append(files)
                for piece_of in self.shit : self.path.remove(piece_of)
                self.dic = dict(sorted(self.dic.items()))

                for item in self.dic :
                    # current name
                    oldname = self.dic[item][0]
                    head = os.path.split(self.dic[item][0])[0] 

                    # split extention of file .
                    ext = os.path.splitext(self.dic[item][0])[1]
                    Time =self.dic[item][1] 
                    newname = head  + '/' + Time + ext

                    # if file with newname is exist , try to change it .
                    while os.path.exists(newname) :
                        if (newname == oldname) : 
                            break
                        newname = head + '/' + Time +'_' + str(self.count) + ext
                        self.count+=1
                    self.count = 1

                    #Renaming...
                    os.rename(self.dic[item][0],newname) 
                    self.cnt+=1

                    #show log
                    if self.Verbos :
                        print ( ' Renamed:' , 
                                self.dic[item][0] , 
                                '->' ,
                                newname.split('/')[-1]
                        )
                if self.Verbos :
                    print (" [" , self.cnt ," : Files Renamed ]")

    # add switchs whit argparse
    def get_parser(self) :

        parser = argparse.ArgumentParser(
                add_help = False ,
                prog = 'Rit',
                usage='Rit [-fjv] [file[strftime]]'
                )

        # verbos
        parser.add_argument (
                '-v',
                '--verbos',
                action = 'store_true'
                )

        # file argument
        parser.add_argument (
                'path',
                type = str,
                nargs = '*'
                )

        # rename all photos in directories recursivly
        parser.add_argument (
                '-r',
                '--recursive',
                action = 'store_true'
                )

        # time format
        parser.add_argument (
                '-f',
                '--format',
                dest = 'format',
                action = 'store_true'
                )

        # jalali time format
        parser.add_argument (
                '-j',
                '--jformat',
                dest = 'jformat',
                action = 'store_true'
                )
        
        # help
        parser.add_argument (
                '-h',
                '-help',
                '--help',
                action = 'store_true'
                )
       
        # version
        parser.add_argument (
                '--version',
                action = 'store_true'
                )

        self.args = vars(parser.parse_args())

        # move path arguments to self.path
        self.path = self.args['path']
        
    # what happen on terminal?! 
    def cli (self) :

        if self.args['verbos'] :
            self.Verbos = True

        if self.args['recursive'] :
            self.recursive = True
        
        if self.args['format'] or self.args['jformat'] :
            if self.args['jformat'] :
                self.jalali = True

            for item in sys.argv :
                #pattern to find timeformat string (%strftime)
                if re.match('^%[^\n]*',item) :
                    self.format = item
                    self.path.remove(item)

        if self.args['help'] :
            print('\nRename It (Version 1.0.2)'
                  '\nusage: Rit [OPTIONS] [FILE|DIR...]'
                '\n\nOptions :'
                  '\n  -v, --verbos      increase output verbosity '
                  '\n  -f, --format      time unix format '
                  '\n  -j, --jformat     jalali time unix format '
                  '\n  -r, --recursive   rename all photos in directories recursivly'
                  '\n  -h, --help        show this help message and exit '
                  '\n  --version         show version '
            )
        
        if self.args['version'] :
            print('''
                          ____  _ _   
                         |  _ \(_) |_ 
                         | |_) | | __|
                         |  _ <| | |_ 
                         |_| \_\_|\__|

                        This is : Rename It
                        Version : 1.0.2
                         Author : Moein Halvaei
                         E-Mail : <moeinn.com@gmail.com>
                      Copyright : (C) 2020  Moein Halvaei
                        License : GNU General Public License, version 3 (gpl3)
                    '''
            )

def main():

    RenameIt = Rename()
    RenameIt.get_parser()  
    RenameIt.cli()
    RenameIt.Rit()

if __name__ == "__main__" :
    main()


